### PR TITLE
Support INTO clauses in AQL hashed COLLECT variant

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Docs PR: https://github.com/arangodb/docs/pull/1295

* Make the hashed variant of AQL COLLECT support INTO clauses too. Previously only the sorted variant of AQL COLLECT supported INTO clauses.

Helps https://arangodb.atlassian.net/browse/PRESUPP-546

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/1295
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/PRESUPP-546
- [ ] Design document:

### CODE REVIEW

**Comment by Simran-B** [comment]
There is at least the following statement in the docs that requires an update with this PR:

`aql/operations-collect.md`
> Not all `COLLECT` statements can use the *hash* method, in particular ones with an `INTO` clause are not eligible.

File: arangod/Aql/HashedCollectExecutor.cpp [hunk 1]
```diff
@@ -191,6 +196,18 @@ void HashedCollectExecutor::writeCurrentGroupToOutput(
...
     }
   }
+  if (_infos.getCollectRegister().value() != RegisterId::maxRegisterId) {
+    auto& builder = *_currentGroup->second.second;
+    builder.close();
+
+    AqlValue val(std::move(*builder.steal()));
+    AqlValueGuard guard{val, true};
+    builder.clear();
+    _currentGroup->second.second.reset();
```
Comment (mpoeter) [inline arangod/Aql/HashedCollectExecutor.cpp:206]: NIT The reset will destroy the builder, so the `builder.clear()` is unnecessary.

File: arangod/Aql/HashedCollectExecutor.cpp [hunk 2]
```diff
@@ -316,6 +333,25 @@ HashedCollectExecutor::findOrEmplaceGroup(InputAqlItemRow& input) {
   auto it = _allGroups.find(_nextGroup);
   if (it != _allGroups.end()) {
     // group already exists
+    if (_infos.getCollectRegister().value() != RegisterId::maxRegisterId) {
+      auto& builder = *((*it).second.second);
```
Comment (mpoeter) [inline arangod/Aql/HashedCollectExecutor.cpp:337]:
```suggestion
      TRI_ASSERT(it->second.second != nullptr);
      auto& builder = *(it->second.second);
```

File: arangod/Aql/HashedCollectExecutor.cpp [hunk 3]
```diff
@@ -355,9 +391,34 @@ HashedCollectExecutor::findOrEmplaceGroup(InputAqlItemRow& input) {
...
+      builder->openObject();
+      for (auto const& pair : _infos.getInputVariables()) {
+        builder->add(VPackValue(pair.first));
+        input.getValue(pair.second)
+            .toVelocyPack(_infos.getVPackOptions(), *builder,
+                          /*resolveExternals*/ false,
+                          /*allowUnindexed*/ false);
+      }
+      builder->close();
+    }
```
Comment (mpoeter) [inline arangod/Aql/HashedCollectExecutor.cpp:415]: Duplication of lines 338-354. Can we extract this into a small helper function?

**Review by mpoeter — APPROVED** [review]

File: arangod/Aql/HashedCollectExecutor.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/Aql/HashedCollectExecutor.cpp]: Done

File: arangod/Aql/HashedCollectExecutor.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/HashedCollectExecutor.cpp]: Done

**Review by mpoeter — APPROVED** [review]

**Comment by jsteemann** [comment]
> There is at least the following statement in the docs that requires an update with this PR:
> 
> `aql/operations-collect.md`
> 
> > Not all `COLLECT` statements can use the _hash_ method, in particular ones with an `INTO` clause are not eligible.

@Simran-B : addressed in https://github.com/arangodb/docs/pull/1295

**Review by mchacki — APPROVED** [review]
LGTM. Thank you

**Pull request merged**